docs/index.rst

```diff
@@ -4,4 +4,315 @@ JDBC Connector
+largest seen so far. "Rewinding" upon failure simply requires looking up the largest ID
+previously seen and starting queries with that value.
+
+However, many tables cannot be guaranteed to only have insertions and have a unique,
+auto-incrementing ID. In order to support a variety of tables with different properties, multiple
+different approaches to defining offsets & queries for new or updated rows are supported.
+
+Incremental Query Modes
+~~~~~~~~~~~~~~~~~~~~~~~
+
+* **Increasing Column**: In this mode, a single column is specified and contains a unique ID for
+  each row which increases with each insertion. Although it could be implemented differently, this
+  type of column is most frequently implemented as an ``AUTOINCREMENT`` column in a table where rows
+  are only inserted, but not updated. (Note that updates will *not* be handled properly in this
+  mode because they cannot be detected once the row has been processed for the first time.)
+
+* **Timestamp Column**: In this mode, a single column is specified and contains a modification
+  timestamp for each row. Using a timestamp makes it simple to query for updated data by simply
+  requesting any rows with newer timestamps than the most recent timestamp previously observed.
+  However, these timestamps are also used as offsets and therefore technically violate the
+  uniqueness requirement for offsets. Because of this, this mode cannot guarantee all updated data
+  will be delivered: if 2 rows share the same offset and are returned by an incremental query, but
+  only one has been processed before a crash, the second update will be missed when the system
+  recovers.
+
+* **Timestamp and Increasing Columns**: This is the most robust and accurate mode, combining an
+  increasing column with a timestamp column. By combining the two, as long as the timestamp is
+  sufficiently granular, each (id, timestamp) tuple will uniquely identify a row update. Even if an
+  update fails midway, when the system recovers, unprocessed updates will still be correctly
+  detected and delivered.
+
+* **Custom Query**: The JDBC connector supports using custom queries instead of copying whole
+  tables. With a custom query, one of the other update automatic update modes can be used as long
+  as the necessary ``WHERE`` clause can be correctly appended (set the appropriate ``mode``
+  configuration). Alternatively, the specified query may handle filtering to new updates itself;
+  however, note that no offset tracking will be performed (unlike the automatic modes where
+  ``increasing`` and/or ``timestamp`` column values are recorded for each record), so the query
+  must track offsets itself.
+
+* **Bulk**: This mode is unfiltered and therefore not incremental at all. It will load all rows
+  from a table on each iteration. This can be useful if you want to periodically dump an entire
+  table where entries are eventually deleted and the downstream system can safely handle duplicates.
```

Comment:
I'd add a note recommending that at least one of the tracking columns will have an index.